apps/studio/src/server/modules/resource/resource.service.ts

```diff
@@ -409,3 +410,183 @@ export const publishResource = async (
 
   return addedVersionResult
 }
+
+export const getWithFullPermalink = async ({
+  resourceId,
+}: {
+  resourceId: string
+}) => {
+  const result = await db
+    .withRecursive("resourcePath", (eb) =>
+      eb
+        .selectFrom("Resource as r")
+        .select([
+          "r.id",
+          "r.title",
+          "r.permalink",
+          "r.parentId",
+          "r.permalink as fullPermalink",
+        ])
+        .where("r.parentId", "is", null)
+        .unionAll(
+          eb
+            .selectFrom("Resource as s")
+            .innerJoin("resourcePath as rp", "s.parentId", "rp.id")
+            .select([
+              "s.id",
+              "s.title",
+              "s.permalink",
+              "s.parentId",
+              sql<string>`CONCAT(rp."fullPermalink", '/', s.permalink)`.as(
+                "fullPermalink",
+              ),
+            ]),
+        ),
+    )
+    .selectFrom("resourcePath as rp")
+    .select(["rp.id", "rp.title", "rp.fullPermalink"])
+    .where("rp.id", "=", resourceId)
+    .executeTakeFirst()
+
+  return result
+}
+
+const getResourcesWithLastUpdatedAt = ({ siteId }: { siteId: number }) => {
+  return db
+    .selectFrom("Resource")
+    .select([
+      "Resource.id",
+      "Resource.title",
+      "Resource.type",
+      "Resource.parentId",
+      // To handle cases where either the resource or the blob is updated
+      sql`GREATEST("Resource"."updatedAt", "Blob"."updatedAt")`.as(
+        "lastUpdatedAt",
+      ),
+    ])
+    .leftJoin("Blob", "Resource.draftBlobId", "Blob.id")
+    .where("Resource.siteId", "=", siteId)
+}
+
+const getResourcesWithFullPermalink = async ({
+  resources,
+}: {
+  resources: Omit<SearchResultResource, "fullPermalink">[]
+}): Promise<SearchResultResource[]> => {
+  return await Promise.all(
+    resources.map(async (resource) => ({
+      ...resource,
+      fullPermalink: await getWithFullPermalink({
+        resourceId: resource.id,
+      }).then((r) => r?.fullPermalink ?? ""),
```

Comment:
question: seems like this is the same steps as using the `getResourceFullPermalink` function above?

Reply:
yup. don't think there's much difference between either other than data structure and `getWithFullPermalink` seem more used by resource side of things

didn't wanna touch `getResourceFullPermalink` - perhaps can add this as a note to refactor to be the same in subsequent PRs?